File: leetcode/array/3595_Once_Twice/codes/bit.py
from typing import List

class Solution:
    def onceTwice(self, nums: List[int]) -> List[int]:
        bits = [0] * 32
        for num in nums:
            x = num & 0xFFFFFFFF
            for i in range(32):
                bits[i] += (x >> i) & 1
        xor_u = 0
        for i in range(32):
            if bits[i] % 3 != 0:
                xor_u |= (1 << i)
        mask = xor_u & -xor_u
        bitsg = [0]*32
        for num in nums:
            x = num & 0xFFFFFFFF
            if (x & mask) != 0:
                for i in range(32):
                    bitsg[i] += (x >> i) & 1
        cand_u = 0
        for i in range(32):
            if bitsg[i] % 3 != 0:
                cand_u |= (1 << i)
        def to_signed(u):
            if u >= (1 << 31):
                return u - (1 << 32)
            return u
        cand = to_signed(cand_u)
        other = to_signed(cand_u ^ xor_u)
        cnt = sum(1 for x in nums if x == cand)
        if cnt == 1:
            return [cand, other]
        else:
            return [other, cand]

File: leetcode/array/3595_Once_Twice/codes/bit.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> onceTwice(vector<int>& nums) {
        long long bits[32] = {0};
        for (int num : nums) {
            uint32_t x = (uint32_t)num;
            for (int i = 0; i < 32; ++i) bits[i] += (x >> i) & 1u;
        }
        uint32_t xor_u = 0;
        for (int i = 0; i < 32; ++i) if (bits[i] % 3 != 0) xor_u |= (1u << i);
        uint32_t mask = xor_u & (~xor_u + 1u);
        long long bitsg[32] = {0};
        for (int num : nums) {
            uint32_t x = (uint32_t)num;
            if (x & mask) for (int i = 0; i < 32; ++i) bitsg[i] += (x >> i) & 1u;
        }
        uint32_t cand_u = 0;
        for (int i = 0; i < 32; ++i) if (bitsg[i] % 3 != 0) cand_u |= (1u << i);
        uint32_t other_u = cand_u ^ xor_u;
        int cand = (int)cand_u;
        int other = (int)other_u;
        int cnt = 0;
        for (int num : nums) if (num == cand) ++cnt;
        if (cnt == 1) return {cand, other};
        else return {other, cand};
    }
};
